### ISSUE TITLE
Use of public DID in connection

### ISSUE THREAD

**Issue by domwoe** [hyperledger/aries-cloudagent-python#358]
I'm wondering what would be the downsides of using the public DID for connections in case of an agent that acts mainly as an issuer. 

I think this could be useful to be implemented as an option to the create and receive invitation endpoints.

**Comment by swcurran** [comment]
That is a viable option that I think is (or should be) included in the 0160-Connections and 0023-DID-Exchange RFCs. Rather then waiting for an invitation, an agent that learns of an issuer's public DID can just send a connection/did-exchange "request" message using the endpoint and keys from the DID. For example, a verifier receiving a verifiable credential issued by an issuer the verifier has never heard up can use the DID linked in the credential definition to connect with and learn more about the issuer.

**Comment by domwoe** [comment]
@swcurran Thank you for your reply.

In ACA-Py it is already possible to use the --public-invites flag. This allows to send a connection request message using the endpoint and keys from the DID. However,  on receiving this message ACA-Py will still create a new pairwise DID for this connection. That's why I'm wondering if there are strong reasons to do this or if there should be an option to use the public DID not only for initiating a connection.

**Comment by swcurran** [comment]
Ah.  That gets into a rabbit hole that has been discussed, but not resolved.

There are a number of people that want to allow an agent to use a public DID for all connections (and a number that don't). The technical challenge right now is that based on the data available to the receiver, there is no way to tell which sender sent the message. The public key of the receiver's DID is used by the receiver to determine the connection. If all the receivers are using the same key, the receiver can't determine the connection.

There are of course ways to solve this. AFAIK, the conversation has never carried on far enough to get to a solution everyone is happy with.

Inputs from others?

**Comment by andrewwhitehead** [comment]
@swcurran ACA-py uses the sender verkey when it's available to identify the connection, so in most cases except for anon-packed messages.

**Comment by swcurran** [comment]
If that's the case, it should work.  Interesting.  There was a discussion on the aries-rfcs about this that AFAIK did not get resolved.  See this issue: https://github.com/hyperledger/aries-rfcs/issues/239

I believe that is the same topic, right?

**Comment by OR13** [comment]
You mean the publicKey thumbprint / fingerprint is used? or the did publicKey.id? or the full publicKey bytes?

the term `public` in front of DID seems unnecessary, these connections are built on the did document data model right?

the did is either resolvable or its not, whether its publicly resolvable is irrelevant no?

**Comment by swcurran** [comment]
@OR13 - there are two ways to use the invitation mechanism to establish a connection starting from the invitation (which is an out-of-band message). One is to have an inline service block to which the next message in the sequence is to be sent.  The second is to provide a public DID (one that is globally resolvable) and a service block from the resolved DID can be used.  The follow up message is a DIDComm message ('connection-request'), and contains (usually) the did:peer DIDDoc that the inviter is to use for the invitee.  Then there is a message ('connection-response') from the inviter that is (usually) the did:peer DIDDoc that the invitee is to use for the inviter.

The "usually" caveats are because in theory, either party could use a public DID in place of the "did:peer" DIDDoc.  We strongly (strongly) recommend against that for people, and there is a friendly split on whether it is a good idea for an organization to use the public DID, with good arguments on both sides. This is the context of this specific issue.

The term "public DID" is often needed in the DIDComm context because of the prevalence of private, "did:peer" DIDs. In Indy at least, the only public DIDs needed are for issuers (they are used in issued credentials) and in the case mentioned above of the organization that uses a public DID instead of "did:peer" in their relationships. Other than those cases, there are not as yet any other obvious uses for public DIDs.

**Comment by domwoe** [comment]
I've created a first PoC of this feature. See https://github.com/boschresearch/aries-cloudagent-python/tree/feature/358-allow-public-did

I've added a `--use_public_did` start parameter. If set, the agent will use the public DID for all connections.
I've added a `use_public_did` query parameter to the `/connections/receive-invitation' route that allows to specify the usage of the public DID for a particular invitation.

So far so good. Now my questions :)

- How should the invitee be able to specify the usage of the public DID?
  - Adding a use_public_did option to the `create-invitation` route would not work for implicit public invitations
  - I think it would be better to allow the decision when a connection request gets accepted. However, currently the peer DID is created when the request is received, not when the request gets accepted.

- Currently I'll send the DID doc with the connection request. Should we send only the DID and the receiver should fetch the information on the ledger?

- Currently you can have multiple connections with another agent. With peer DIDs that's not an issue, but by using the public DID this gets messy. I'm not sure how to handle this. Should we raise an error when there is already a connection record with the same DIDs or should we update the old record?

**Comment by domwoe** [comment]
Concerning the first Question:

Let's distinguish between the two roles in the connection protocol

**Invitee**

This is straight forward from my point of view with the approach described above. 
In `/connections/receive-invitation` I can specify the usage of the public did by setting the `use_public_did` query parameter accordingly

**Inviter**

Given the current implementation of the connection protocol, there is no way to uniquely identify an invitation besides the peer DID/verkey used. This means as an inviter I can only use the public DID for global public connections, but I cannot use the public DID for "private connections".

**Comment by domwoe** [comment]
@swcurran @sklump  Would be great to get your feedback/opinion on this

**Comment by domwoe** [comment]
The way "DIDcomm" works right now, there seem to be a big drawback in using public DIDs for connections. 
Please correct me if I'm wrong, but it seems that if an attacker would record all communications based on that DID and could get hands on the private key at some point, the attacker would be able to decrypt all messages.

So maybe there needs to be  a way to establish a connection based on the public DID but then switch to a peer DID. Or a standard way to prove control of the public DID in an existing pairwise connection.

**Comment by swcurran** [comment]
DID Exchange (RFC 0023) just had a clarification to do that. Although a public DID may be used to start a connection (in implicit invitation), an agent may choose to create the connection using a peer DID and not use the public DID thereafter. 

The pros and cons of public vs. peer DID for connections:

- Public has only a key pair to manage -- a key rotation published to a ledger changes it for all contacts
   - Of course that leads to a caching and TTL problem for the other side of the connection...
- With peer DIDs there are many key pairs to manage -- one per connection
   - But so many keys may be hard for enterprises to manage

I've heard on the DIDComm V2 calls (at DIF) some discussion of Perfect Forward Secrecy (where getting a private key does not enable decryption of all past messages), but I don't think the issue has been settled.

**Comment by domwoe** [comment]
I haven't looked into DID Exchange until now that I've noticed that @sklump has implemented (parts of?) the protocol.

However, I'm not sure how DID Exchange solves my problem :) 

Given the roles in RFC 0023.
As a *responder* I can do what you describe. I publish my public DID. The *requester* used the public DID as an implicit invitation and creates a request. The *responder* creates a new DID and sends a response. Done.

However, in the role of a *requester*, how can I accomplish the same?
* I want to communicate to the responder that I'm controlling a specific public DID
* Afterwards I may want to use new DID for the communication.

Furthermore, given we use an implicit invitation. Does the connection record keep the initial public DID and the actual peer DID for communication?

**Comment by marc0olo** [comment]
@domwoe what I am trying to achieve now is to provide a public DID document for an ACA-Py instance which can be used by an aries-go based agent to create and accept an implicit invitation (RFC 0023) from the information in the DID document

if I get this working I can give you an answer about the last question. my guess is that the connection record will only contain the peer DIDs.

**Comment by domwoe** [comment]
@marc0olo I'm not sure how aries-go handles this, but the idea of an implicit invitation is that you don't need an invitation to start the flow.

With ACA-Py and the connections protocol you (as the controller) would post 
```
{
did: did:sov.... // the public DID
}
```

to the `/receive-invitation` endpoint.

(I'm writing this from memory, so there might be some errors, but that's the idea at least).

ACA-Py will resolve the DID document, retrieve the DIDcomm endpoint and the necessary keys and create a connection request.

**Comment by marc0olo** [comment]
@domwoe I am currently only thinking of the RFC 0023 way. and there are two [different approaches](https://github.com/hyperledger/aries-rfcs/tree/master/features/0023-did-exchange#correlating-requests-to-invitations)  for the invitation part:
- explicit invitation via [RFC 0434](https://github.com/hyperledger/aries-rfcs/blob/master/features/0434-outofband/README.md#messages) (out-of-band)
   - for this type of invitation aries-go and ACA-Py are not compatible right now from what I analyzed so far. but it works and can be used between two different ACA-Py agents
- implicit invitation which is included in the service attribute of a public DID document and should from my understand be conform to [RFC 0067](https://github.com/hyperledger/aries-rfcs/blob/master/features/0067-didcomm-diddoc-conventions) (didcomm-diddoc-conventions)
  - this is what I am talking about. in **aries-go** the there exists a controller binding which is called `create-implicit-invitation` where one can provide a public DID which will be used to start the DIDExchange and returns a `connectionId`

I am not sure whether the implicit invitation between ACA-Py and aries-go can be established this way at this point of time. but I am looking into it and hope that I can make use of it as long as the did document follows the conventions of RFC 0067 (assuming that aries-go expects these conventions).

Is that what you are looking for? (if you leave out the topic with aries-go for a moment)

When I am right the "creation" (or "lookup") of an implicit invitation is currently not implemented in ACA-Py. This is why I tagged @sklump in my comment above.

**Comment by jpbu** [comment]
@domwoe your tests look good, thanks. How exactly did you create the DID and set the `did-communication` service endpoint required for the implicit invitation?

Given the linked issue https://github.com/hyperledger-labs/business-partner-agent/issues/300 I tried to register an Indy DID (`did:sov:iil`) manually and set the service endpoint with an `ATTRIB` transaction using the Indy CLI. I was able to set an attribute like `{"endpoint":{"did-communication":"http://xy.z:1234"}}` which is then displayed in the resolved DID document as `"service":{"type":"did-communication","serviceEndpoint":"http://xy.z:1234"}` (again, no array, but I guess that is the same resolver issue mentioned before). I did not find out how to add further attributes for the service like the id (which is mandatory according to the W3C spec) or the recipientKeys (which are required/expected for the implicit invitation in the DID Exchange).

I looks to me like the missing further attributes for the service in the DID document are a general limitation in Indy or at least this would need to be addressed in the Indy SDK and CLI first - do you agree?

**Comment by domwoe** [comment]
@jpbu 

> I looks to me like the missing further attributes for the service in the DID document are a general limitation in Indy or at least this would need to be addressed in the Indy SDK and CLI first - do you agree?

it's not so much about what's stored on the ledger but how that gets interpreted by the resolver. Have a look at the implementation of Scoir I linked above. They generate a compliant did doc that Aries Framework Go can consume from a `NYM` an `ATTRIB` tx.  

The general issue is that `did:sov` is not well defined and much older than the current did spec. Many did drivers in the universal resolver are not compliant with the current did spec. The implementation of `did:sov` is one example. However, the did spec is now in review and there's an ongoing effort to define a compliant `did:indy` method.

**Comment by swcurran** [comment]
Regards `did:sov` being old.  We are currently hard at working spec'ing out `did:indy` that will replace `did:sov` with lots of new capabilities - most importantly network namespaces and standard DIDDocs from the ledger.  Glad to get help on that if you are interested.

**Comment by marc0olo** [comment]
In regards to the implicit invitation I opened following issues:
- https://github.com/hyperledger/aries-cloudagent-python/issues/881
- https://github.com/hyperledger/aries-framework-go/issues/2415

@swcurran thanks for the information. can you provide us links to the WIP state of the `did:indy` spec?

**Comment by swcurran** [comment]
The meetings and details can be found here: https://wiki.hyperledger.org/display/indy/Indy+DID+Method+Specification

The Hack MD document needs an update -- should happen over the Holidays.  The details of the work are in the meeting minutes from the last week's meeting.

**Comment by SalimiHabib** [comment]
It seems go framework already have  implementation of this issue as [future ](https://github.com/sandrask/aries-framework-go/commit/5d7483f50ab098838880202959ef9b1ca65549ce?branch=5d7483f50ab098838880202959ef9b1ca65549ce&diff=unified) in client library

**Comment by domwoe** [comment]
I'm just reviewing this issue in light of the progress on DID Exchange. 
I still think DID Exchange as implemented right now does not address this issue.

I can use 
```
​/didexchange​/create-request
```
to create and send and a DID Ex request by providing a public DID of the responder, but I don't see a way to tell the responder that I (the requester) can be addressed by a public DID.

@sklump  Is this correct?

Could we add a property `use_public_did` to `​/didexchange​/create-request` 
such that `did: public_did` and no `did_doc~attach` is provided in the request message?

**Comment by domwoe** [comment]
Implemented in PR #1220
